### ISSUE TITLE
document the config file

### ISSUE THREAD

**Issue by cjyetman** [RMI-PACTA/workflow.data.preparation#7]
in the README?

**Comment by jdhoffa** [comment]
Agree with this, though I think documenting a `.yml` in the README is probably  a bit cumbersome. 
Found this article:

**Comment by cjyetman** [comment]
also [YAML comments](https://www.educative.io/answers/how-to-write-comments-in-yaml?utm_campaign=brand_educative&utm_source=google&utm_medium=ppc&utm_content=performance_max&eid=5082902844932096&utm_term=&utm_campaign=%5BNew%5D+Performance+Max&utm_source=adwords&utm_medium=ppc&hsa_acc=5451446008&hsa_cam=18511913007&hsa_grp=&hsa_ad=&hsa_src=x&hsa_tgt=&hsa_kw=&hsa_mt=&hsa_net=adwords&hsa_ver=3&gclid=CjwKCAjwzuqgBhAcEiwAdj5dRnfdXAls2gvLJnD_FDtWA4mguy_-mby3mIh_PJiHRZvUfEcBxOckkRoCwf8QAvD_BwE)?

**Comment by jdhoffa** [comment]
I saw this comment in the post I linked... honestly not totally sure what it means, but seems to be swaying away from comments:
``` yml
  ## comments
  ## NOTE: YAML generally throws your comments away, so they are not very useful
  ## for round-trip metadata
```

**Comment by AlexAxthelm** [comment]
TL;DR: YAML comments would be fine, but probably not comprehensive. Switching to a JSON config file(s) would be a big change (conceptually, probably not that bad in-code), but could give us access to better tooling.

----
Some of the issues I see here (that I don't necessarily have excellent solutions for with our current setup):

1. We aren't clearly identifying required vs optional parameters (what causes data prep to fail or not)
2. What parameters are permitted (are there params in the file that aren't being used? Example here would be `update_factset` (at the moment)).
3. Required types aren't clear (what is supposed to be a string, int, bool, list, etc)
4. Acceptable values aren't clear (what are valid values to put in `sector_list`?)
5. What is the description of each field?
6. Is there a good way to keep the actual config in sync with the docs?

As I see it, the biggest thing working against us here is `{config}` package. It does have the nice feature of defining the `inherits:` key, allowing us to not re-specify the entire config every time (effectively only specifying the diffs between configs), but it only works with YAML, while JSON has JSON Schema which addresses all of the points above (I haven't found a good equivalent for YAML).

Additionally, there are some tools (I haven't tested them) to turn a JSON schema into human documentation: https://github.com/coveooss/json-schema-for-humans

here's a (very limited) example of what that might look like:
```json
{
  "$schema": "http://json-schema.org/draft-07/schema",
  "title": "portfolioMetadata",
  "description": "Portfolio reexported form workflow.portfolio.parsing",
  "type": "object",
  "properties": {
    "export_sqlite_files": {
      "description": "Should PACTA-data be exported to SQLite files?",
      "type": "boolean"
    },
    "imf_quarter_timestamp": {
      "description": "Timestamp of the IMF quarter used for emissions data",
      "type": "string",
      "pattern": "20[0-9]{2}-Q[1-4]"
    },
    "scenario_sources_list": {
      "description": "List of scenario sources to be used for dataset",
      "type": "array",
      "items": {
        "enum": [
          "ETP2020",
          "GECO2021",
          "GECO2022",
          "IPR2021",
          "ISF2021",
          "WEO2021",
          "WEO2022"
        ]
      }
    
    }
  },
  "required": ["export_sqlite_files", "imf_quarter_timestamp", "scenario_sources_list"]
}
```

This schema:
* requires that `export_sqlite_files be boolean
* `imf_quarter_timestamp` be a string matching a regex
* `scenario_sourcces_list` be an array (R: vector) of values belonging to a predefined list
* Provide descriptions for all those properties
* mark them all as required

the `jsonvalidate` package works in R to take a json string, and check it against a schema like this, and then we can unpack it to an R list.

----

I don't think converting to JSON would be a huge burden after we import the config (they both convert to `lIst` objects), so the main question is if we would want to keep inheritance or not. The primary downside of losing that inheritance is that if we need to change any properties, propagating them would be on us (manual, and therefore fallible).

If we're okay with losing inheritance, then the process is pretty straightforward:
1. Define our schema
2. make JSON files with all the required keys, and appropriate values
3. change the call to `config::get()` to something like
```r
config_json <- file.path("wherever/we/put/these/", paste0(Sys.getenv("R_CONFIG_ACTIVE"), ".json"))
jsonvalidate::schema$new("some/path/schema.json")$validate(config_json)
config <- read_json(config_json, simplify_vector = FALSE)
# continue as normal
```

If we want to keep inheritance, then things get a little more tricky, but not too bad. The sketch here looks something like:
```json
{
  "default": {
    "export_sqlite_files": false,
    "imf_quarter_timestamp": "2021-Q2",
    "scenario_sources_list": [
      "GECO2021",
      "IPR2021",
      "ISF2021",
      "WEO2021",
    ]
  },
  "2022Q4": {
    "inherits": "default",
    "scenario_sources_list": [
      "GECO2022",
      "WEO2022"
    ]
  },
  "2022Q4-SQLite": {
    "inherits": "2022Q4",
    "export_sqlite_files": true
  }
}
```

```r
all_configs <- jsonlite::read_json("configs.json")
active_config_list <- all_configs[[Sys.getenv("R_ACTIVE_CONFIG", "default")]]
if ("inherits" %in% names(active_config)){
  # steal the config::merge function, and let it run recursively until we get back to default
  active_config_list <- merge(all_configs[[active_config_list[["inherits"]] ]], active_config_list)
}

jsonvalidate::schema$new("some/path/schema.json")$serialise(active_config_list)
# ^^^ This actually turns the list object back to a JSON string, but it does fail if it can't do that correctly, so it's still a good check.
config <- active_config_list
# continue as normal
```

Anyways, YAML comments work as a first pass on this, but if we want to do anything more formal, we may want to seek out a different system

**Comment by jdhoffa** [comment]
Seems like a very good candidate for a tech review. 

As usual, I see a decent number of pros and cons with either approach, and don't have a clear personal preference (when that is the case for me, I usually would fall back on "leave it how it is"). 

Better and more type-safe (ish) documentation seems like a big win though.  

Let's decide as a team?

**Comment by AlexAxthelm** [comment]
@jdhoffa FYI, I already laid out a similar proposal for converting the portfolio CSV+YAML to JSON at a recent tech review, with a result of "we can convert the portfolio parameter YAML to JSON (for webapp), but let's not touch the CSV for now".

**Comment by jdhoffa** [comment]
Gotcha, then maybe let's do it there first? Prototype it, document it, and we can have a working demo to see how it works. And if it seems like a good thing, we can update the files here too?

**Comment by cjyetman** [comment]
This repo uses a YAML config file, and it has significant infrastructure built around reading that in, managing cascading options, etc. so I do not intend to change that. Eventually, I will document the parameters/options, likely both in the [config.yml](https://github.com/RMI-PACTA/workflow.data.preparation/blob/main/config.yml) file and in the README.

**Comment by jdhoffa** [comment]
Tech review topic added [here](https://loop.microsoft.com/p/eyJ3Ijp7InUiOiJodHRwczovL3JvY2ttdG5pbnMuc2hhcmVwb2ludC5jb20vP25hdj1jejBsTWtZbVpEMWlJVkl0UTNkRlRFTmFNVVUyTVdVdFlXSmxNSGMyUldkTmVVNHRSSGhYZDJoRGRHcDJhakF3ZDBwNFQxWklhRzUwYmxaamJVUlRjV1Z0VFZBMmIyOW1WemNtWmowd01WcERXa1pVTlZGSE5VRkdNMGhDUWtoVlRrRkxOemRhTWxGR1dUWTBXa0UzSm1NOUptWnNkV2xrUFRFJTNEIiwiciI6ZmFsc2V9LCJwIjp7InUiOiJodHRwczovL3JvY2ttdG5pbnMuc2hhcmVwb2ludC5jb20vOmZsOi9yL2NvbnRlbnRzdG9yYWdlL0NTUF8xMGIwZTA0Ny05OWIwLTRlZDQtYjU3Yi1lNjliN2I0YzNhMTIvRG9jdW1lbnQlMjBMaWJyYXJ5L0xvb3BBcHBEYXRhL1VudGl0bGVkLmxvb3A%2FZD13MGVmZTNkOTQzZGZkNDM3NGFkM2YxOTdkNGJiMjk0NTUmY3NmPTEmd2ViPTEmbmF2PWN6MGxNa1pqYjI1MFpXNTBjM1J2Y21GblpTVXlSa05UVUY4eE1HSXdaVEEwTnkwNU9XSXdMVFJsWkRRdFlqVTNZaTFsTmpsaU4ySTBZek5oTVRJbVpEMWlJVkl0UTNkRlRFTmFNVVUyTVdVdFlXSmxNSGMyUldkTmVVNHRSSGhYZDJoRGRHcDJhakF3ZDBwNFQxWklhRzUwYmxaamJVUlRjV1Z0VFZBMmIyOW1WemNtWmowd01WcERXa1pVTlZWVlNGZzNRVFUzU2pWUFVrSXlNbEJaV2xCV1JqTkdSa05XSm1NOUpUSkdKbVpzZFdsa1BURW1ZVDFNYjI5d1FYQndKbkE5SlRRd1pteDFhV1I0SlRKR2JHOXZjQzF3WVdkbExXTnZiblJoYVc1bGNpWjRQU1UzUWlVeU1uY2xNaklsTTBFbE1qSlVNRkpVVlVoNGVXSXlUbkppV0ZKMVlWYzFla3h1VG05WldFcHNZMGM1Y0dKdVVYVlpNamwwWmtkSmFGVnBNVVJrTUZaTlVURnZlRkpVV1hoYVV6Rm9XVzFWZDJSNldrWmFNREUxVkdreFJXVkdaRE5oUlU0d1lXNWFjVTFFUWpOVGJtaFFWbXRvYjJKdVVuVldiVTUwVWtaT2VGcFhNVTVWUkZwMllqSmFXRTR6ZDNkTlZuQkVWMnRhVlU1V1JraE9WVVpIVFRCb1ExRnJhRlpVYTBaTVRucGtZVTFzUmtkWFZGa3dWMnRGTXlVeU1pVXlReVV5TW1rbE1qSWxNMEVsTWpJeU1UUTBaR1UyWmkweFkyVTNMVFJqT1RFdE9EY3lNUzB6TjJSa056VTRZMkV4TWpRbE1qSWxOMFElM0QiLCJyIjpmYWxzZX0sImkiOnsiaSI6IjIxNDRkZTZmLTFjZTctNGM5MS04NzIxLTM3ZGQ3NThjYTEyNCJ9fQ%3D%3D)

**Comment by cjyetman** [comment]
closed by #162
